### PR TITLE
DOC-1132 log sink buffering

### PR DESCRIPTION
**Pull request by andf-crl**
Addresses: DOC-1132, DOC-4651

- Adds `buffering` conf option to all 4 supported log sink targets, with reference and usage in `configure-logs.md` and example usage in `logging-use-cases.md`.
- Tried to reconcile semi-conflicting content from DOC-4651, but will need your help to clarify, please see questions.

QUESTIONS:
- This feature is described as available for all log sink targets, and includes example code using `file-groups`, but includes an explicit error message advising that file-groups are not supported: `unimplemented: "buffering" not yet supported for file-groups` (appears in `pkg/util/log/logconfig/validate.go`). It appears to link to #[72452](https://github.com/cockroachdb/cockroach/issues/72452). Which is the case?
- What to do with the existing `buffer-writes` log configuration option? How do `buffering` and `buffer-writes` interact? Raph also asked this [here](https://github.com/cockroachdb/cockroach/pull/70330#issuecomment-959348775). Example: `auditable: true` disables `buffered-writes` if the sink is under `file-groups`. Is this also true for `buffering`?
- Just wanted to confirm that whichever of `max-staleness` or `flush-trigger-size` is hit first will trigger the flush. Think this is the case (and have documented as such), but this Q to verify.

NEW QUESTIONS FROM DOC-4651
- DOC-1132 includes `max-in-flight` sub-param to `buffering`, but DOC-4651 do not include. Similarly, DOC-4651 includes `max-buffer-size` sub-param to `buffering`, but DOC-1132 does not. Guess: `max-in-flight` is gone in favor of `max-buffer-size`, which makes 1000% more sense from a db operator perspective. Please confirm so I can happily delete the confusing `max-in-flight` param from my copy with gusto. I do not see it referenced in `debug check-log-config` output, so I have my delete key ready.
- What is default to `max-in-flight`, if it still exists?

STAGING:
- [configure-logs.md](https://deploy-preview-15311--cockroachdb-docs.netlify.app/docs/dev/configure-logs.html)
- [logging-use-cases.md](https://deploy-preview-15311--cockroachdb-docs.netlify.app/docs/v22.2/logging-use-cases#network-logging-with-log-buffering)

### CODE REVIEW

**Comment by andf-crl** [comment]
@abarganier passing to you for a tech review, as the original eng is no longer at CRL (and you have done related work), but feel free to pass along if you know of a better review target! Perhaps Raph, who reviewed the originating eng PR? TIA either way!!

**Comment by abarganier** [comment]
Happy to take a look! I'll address some questions below.

> This feature is described as available for all log sink targets, and includes example code using file-groups, but includes an explicit error message advising that file-groups are not supported: unimplemented: "buffering" not yet supported for file-groups (appears in pkg/util/log/logconfig/validate.go). It appears to link to #https://github.com/cockroachdb/cockroach/issues/72452. Which is the case?

We're in a state right now where the buffering system used by file-group log sinks differs from the buffering system used by fluent-servers and http-servers. 

file-group sinks therefore don't use the `buffering` config attribute. Instead, they are still on the old `buffered-writes` boolean setting. The newer `buffering` config attribute is only available for fluent-servers and http-servers. Therefore, https://github.com/cockroachdb/cockroach/issues/72452 is to track the work involved to unify the two systems. 

Note that stderr doesn't explicitly support buffering of *any* type.

> What to do with the existing buffer-writes log configuration option? How do buffering and buffer-writes interact?

Given the previous Q&A, I think we can just keep the config options separate for now. They aren't expected to interact - fluent-servers/http-servers will only respect `buffering`, and file-groups will only respect `buffered-writes`.

> auditable: true disables buffered-writes if the sink is under file-groups. Is this also true for buffering?

Great question (one I had to do some code investigation to figure out)! 

For a fluent-server, if `auditable: true`, then `exit-on-error` is [set to true for the fluent-server](https://github.com/cockroachdb/cockroach/blob/967f8741673a89ea598d7eb38f42edee84d0d636/pkg/util/log/logconfig/validate.go#L438-L442). This doesn't disable `buffering`, but if the buffer encounters an error when flushing logs to the fluent server, it will stop the Cockroach node.

The http-server, unfortunately, does no such thing. Setting `auditable` true/false doesn't impact `exit-on-error` or `buffering` in any way. 

This is a confusing inconsistency 😞 I've created https://github.com/cockroachdb/cockroach/issues/90549 to address so we can reconcile this behavior. 

> Just wanted to confirm that whichever of max-staleness or flush-trigger-size is hit first will trigger the flush. Think this is the case (and have documented as such), but this Q to verify.

Correct! max-staleness and flush-trigger-size control the circumstances under which the buffer sink automatically flushes its contents to the target. Zero values disable these flush triggers.

--

I'm going to post a separate comment to answer the questions in the "NEW QUESTIONS..." section from the original post.

**Comment by abarganier** [comment]
Some more Q&A:

> [DOC-1132](https://cockroachlabs.atlassian.net/browse/DOC-1132) includes max-in-flight sub-param to buffering, but [DOC-4651](https://cockroachlabs.atlassian.net/browse/DOC-4651) / [DOC-5728](https://cockroachlabs.atlassian.net/browse/DOC-5728) do not include. Similarly, [DOC-4651](https://cockroachlabs.atlassian.net/browse/DOC-4651) includes max-buffer-size sub-param to buffering, but [DOC-1132](https://cockroachlabs.atlassian.net/browse/DOC-1132) does not. Guess: max-in-flight is gone in favor of max-buffer-size, which makes 1000% more sense from a db operator perspective. Please confirm so I can happily delete the confusing max-in-flight param from my copy with gusto. I do not see it referenced in debug check-log-config output, so I have my delete key ready.

You've got it! `max-in-flight` was deprecated in favor of `max-buffer-size`, which is more descriptive as you've noted. To clarify, the available `buffering` options are (for example):
```
    buffering:
      max-staleness: 5s	
      flush-trigger-size: 1mib
      max-buffer-size: 50mib
``` 

> What is default to max-in-flight, if it still exists?

While it doesn't exist anymore, its replacement, `max-buffer-size`, has a default of `50mib`. 

---

I hope this helps! Let me know if you have any further questions.

I'll go ahead and review the actual docs changes now 🙂

File: v22.2/configure-logs.md [hunk 1]
```diff
@@ -112,6 +113,7 @@ All supported sink types use the following common sink parameters:
...
... in v22.2** Configures buffering of log messages for the sink, with the following sub-parameters:<br><br><ul><li>`max-staleness`: The maximum time a log message will wait in the buffer before a flush is triggered. Default: `5s`</li><li>`flush-trigger-size`: The number of bytes that will trigger the buffer to flush. Default: `1MiB`</li><li>`max-buffer-size`: The maximum size of the buffer: new log messages received when the buffer is full cause the older messages to be dropped. Default: `50MiB`</li><li>`max-in-flight`: The maximum number of concurrent active buffered flushes before messages start being dropped. Default: `??`</li></ul>`buffering` is enabled by default for [Fluentd-compatible](#output-to-fluentd-compatible-network-collectors) and [HTTP](#output-to-http-network-collectors) log sinks, and disabled by default for [log files](#output-to-files) and the [standard error stream (`stderr`)](#output-to-stderr) log sinks.<br><br>To explicitly disable log buffering, specify `buffering: NONE` instead. This setting is typically disabled for [security-related logs](logging-use-cases.html#security-and-audit-monitoring). See [Log buffering](#log-buffering) for more details and usage.|
```
Comment (abarganier) [inline v22.2/configure-logs.md:116]: > disabled by default for [log files](#output-to-files)

As discussed in the Q&A, I believe buffering as a concept is enabled by default for file sinks, but is instead covered under the `buffered-writes` config option.

File: v22.2/configure-logs.md [hunk 2]
```diff
@@ -536,6 +538,54 @@ sinks:
 To ensure that you are protecting sensitive information, also [redact your logs](#redact-logs).
 {{site.data.alerts.end}}
 
+## Log buffering
+
+{% include_cached new-in.html version="v22.2" %} All CockroachDB [log sinks](#configure-log-sinks) support the buffering of log messages, and buffering is enabled by default for the [Fluentd-compatible](#output-to-fluentd-compatible-network-collectors) and [HTTP](#output-to-http-network-collectors) log sinks.Previous to version v22.2, only the [log file](#output-to-files) log sink supported log message buffering.
```
Comment (abarganier) [inline v22.2/configure-logs.md:543]: > All CockroachDB [log sinks]

nit: except for `stderr` AFAIK

File: v22.2/configure-logs.md [hunk 3]
```diff
@@ -536,6 +538,54 @@ sinks:
...
+
+With log buffering configured, log messages are held in a buffer for a configurable time period or accumulated message size before being written to the target log sink together as a batch. Log buffering helps to ensure consistent low-latency log message writes over the network even in high-traffic, high-contention scenarios.
+
+The following shows a basic log configuration with buffering configured for [log file](#output-to-files) writes:
+
+~~~ yaml
+file-defaults:
+  dir: logs
+  buffering:
```
Comment (abarganier) [inline v22.2/configure-logs.md:552]: nit: as discussed in the Q&A, `file-defaults` would use `buffered-writes` instead.

File: v22.2/configure-logs.md [hunk 4]
```diff
@@ -536,6 +538,54 @@ sinks:
...
+    max-staleness: 20s
+    flush-trigger-size: 2MiB
+    max-in-flight: 10
+sinks:
+  file-groups:
+    health:
+      channels: HEALTH
+      buffering:
```
Comment (abarganier) [inline v22.2/configure-logs.md:560]: ditto here. Maybe we should swap with a `fluent-server` instead for example's sake?

File: v22.2/logging-use-cases.md [hunk 5]
```diff
@@ -452,6 +452,40 @@ For example, this JSON message found in the `OPS` logging channel contains a [`n
...
+For example, the following logging configuration adjusts the default log buffering behavior for both a [Fluentd-compatible](configure-logs.html#output-to-fluentd-compatible-network-collectors) and an [HTTP](configure-logs.html#output-to-http-network-collectors) log sink destination:
+
+~~~ yaml
+fluent-defaults:
+  buffering:
+    flush-trigger-size: 2MiB
+http-defaults:
+  buffering:
+    max-staleness: 10s
+    max-in-flight: 10
```
Comment (abarganier) [inline v22.2/logging-use-cases.md:470]: nit: just noting the cleanup here from the Q&A - we can drop `max-in-flight` 🙂

**Review by abarganier — COMMENTED** [review]
Some initial feedback on the content. I know a lot of this will be cleaned up based on the above Q&A, though!

**Review by andf-crl — COMMENTED** [review]
Many thanks for the fantastically-helpful feedback!! 🙂

- I've split `buffering` from all-applicable param list to the specific sinks it is supported for: `fluentd` and `http`. Did not touch existing `buffer-writes` param then, as it still applies in full to `file-groups`, and is unrelated to `buffering` except in concept.
- Removed mention of previous buffering behavior from `fluentd` and `http` sections, as the new `buffering` defaults introduce replacement behavior.
- No longer discussing `stderr` in any fashion in this PR, as `buffering` does not apply to this sink.
- Deleted `max-in-flight` and replaced with `max-buffer-size`. Updated defaults.
- Left `auditible: true` as is for now. Unsure if there is an appropriate place to doc this, given that it does not interact with buffering per se (does not disable) just ... hangs the node. If you have an idea where this belongs, or how to communicate, please file a new docs ticket and I'll make the update!

Thanks again!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/15311#-:-NFAZrZs6RfdfQCiU9WV:bqbiaub)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier)

___
*[`v22.2/configure-logs.md` line 116 at r2](https://reviewable.io/reviews/cockroachdb/docs/15311#-NDioN_bwGXvg8ZB7Osq-r2-116:-NFAZQRU4wopcpovmrTH:bptlgma) ([raw file](https://github.com/cockroachdb/docs/blob/31c6ef3daadf27f964ddf0eb6bad93e752c024df/v22.2/configure-logs.md#L116)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

> disabled by default for [log files](#output-to-files)

As discussed in the Q&A, I believe buffering as a concept is enabled by default for file sinks, but is instead covered under the `buffered-writes` config option. 
</blockquote></details>

TY! I've moved `buffering` to specific param lists for `fluentd` and `http` sinks, now that I know it's no longer applicable in kind to all long sinks.
___
*[`v22.2/configure-logs.md` line 552 at r2](https://reviewable.io/reviews/cockroachdb/docs/15311#-NDioN_bwGXvg8ZB7Osq-r2-552:-NFAZiD6A4Kv5nBLAkXF:b-b6al6n) ([raw file](https://github.com/cockroachdb/docs/blob/31c6ef3daadf27f964ddf0eb6bad93e752c024df/v22.2/configure-logs.md#L552)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: as discussed in the Q&A, `file-defaults` would use `buffered-writes` instead. 
</blockquote></details>

Switched to fluentd example, per your next comment. Right idea!
___
*[`v22.2/logging-use-cases.md` line 470 at r2](https://reviewable.io/reviews/cockroachdb/docs/15311#-NDioN_bwGXvg8ZB7Osr-r2-470:-NFAZnQL72Tq0BuvepYD:b-87j1r7) ([raw file](https://github.com/cockroachdb/docs/blob/31c6ef3daadf27f964ddf0eb6bad93e752c024df/v22.2/logging-use-cases.md#L470)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: just noting the cleanup here from the Q&A - we can drop `max-in-flight` 🙂 
</blockquote></details>

Gone!

**Review by abarganier — APPROVED** [review]
Looks great, :lgtm_strong: 

Awesome work @andf-crl! Thanks for the crystal clear docs 💎 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/15311#-:-NFEbuRIBgzUXUcnefPz:b-75i25i)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andf-crl)

**Comment by andf-crl** [comment]
TYFR @abarganier !

**Review by MichaelTrestman — APPROVED** [review]
LGTM

**Comment by andf-crl** [comment]
TYFR!

**Pull request merged**